### PR TITLE
[kotlin/en]  [In progress] Add Kotlin-lang

### PR DESCRIPTION
**Pull request by loolooyyyy**
I would definitely complete it but before it happens, it would be nice to have someone take a look at it. Some parts are probably unnecessary and unfit for learnxinyminutes. Bad English and grammar mistakes must be checked too :D

There wasn't anything mentioned about this on learnxinyminutes-docs: Should an article be fully authored before committing to main branch, or is it okay to commit it now?

### CODE REVIEW

File: kotlin.html.markdown [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+```java
+
+// Single-line comments start with //
+
+/*
+Multi-line comments look like this.
+*/
+
+// Package specification should be at the top of the source file, but need not
+// to match the directory structure.
```
Comment (yole) [inline kotlin.html.markdown]: need not match

File: kotlin.html.markdown [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+package my.demo
+
+// Importing packages from java
+import java.util.ArrayList;
```
Comment (yole) [inline kotlin.html.markdown]: no semicolons please

File: kotlin.html.markdown [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+          + "\nHex 0x100:\t" + 0x100
+          + "\nLong value, denoted with capital L 100L:\t" + 100L
+          + "\nBinary 0b00000100:\t" + 0b00000100
+          + "\nOctal literal not supported!"
+          + "\nFloating-point, Double by default 123.5:\t" + 123.5
+          + "\nFloating-point, Double by default 123.5e10:\t" + 123.5e10
+          + "\nFloat tagged with F 123.5f:\t" + 123.5f
+  )
+
+  // String literal, denoted by double quote. String are iterable
```
Comment (yole) [inline kotlin.html.markdown]: "double quotes"; "strings are iterable"

File: kotlin.html.markdown [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+  // Local read-only variable.
+  val a: Int = 1;
+  val b = 1
+  val c: Int // Type required when no initializer is provided
+  c = 1
+  
+  // The following line generates compiler error
+  // a += 1
+
+  // Local mutable value, it's type is inferred.
```
Comment (yole) [inline kotlin.html.markdown]: its

File: kotlin.html.markdown [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+  // Read-only map
+  val map = mapOf("a" to 1, "b" to 2, "c" to 3)
+  println(map["s"])
+}
+
+// _________________________________________________________________ FUNCTIONS
+
+// A function with no args, and return type not specified.
+fun funDemo() {
+  return 42;
```
Comment (yole) [inline kotlin.html.markdown]: no semicolons please

File: kotlin.html.markdown [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+}
+
+// Function with both parameter and return type Int.
+fun funDemoSum(a: Int, b: Int): Int {
+  return a + b
+}
+
+// Return type automatically inferred. But such fuction is not visible outside
```
Comment (yole) [inline kotlin.html.markdown]: "such a function"

File: kotlin.html.markdown [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+/**
+ * Any expression in Kotlin may be marked with a label. Labels have the form 
+ * of the @ sign followed by an identifier, for example: @abc, @fooBar
+ */
+fun breakWitLabels() {
+  @loop for (i in 1..100) {
+    for (j in 1..100) {
+      if (...)
+        // jumps to the execution point right after the loop marked with the 
+        // label. A continue would instead proceeds to the next iteration of
```
Comment (yole) [inline kotlin.html.markdown]: "would instead proceed"

File: kotlin.html.markdown [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+// _________________________________________________________ RETURNS AND jUMPS
+
+/**
+ * Any expression in Kotlin may be marked with a label. Labels have the form 
+ * of the @ sign followed by an identifier, for example: @abc, @fooBar
+ */
+fun breakWitLabels() {
+  @loop for (i in 1..100) {
+    for (j in 1..100) {
+      if (...)
```
Comment (yole) [inline kotlin.html.markdown]: put a real condition in here?

File: kotlin.html.markdown [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+}
+
+// Automatic cast for right-hand side of `&&`
+fun nullChecking2(obj: Any): Int? {
+  if (obj is String && obj.length > 0)
+    return obj.length
+}
+
+// *If not null* shorthand 
+fun nullChecking3(obj: Any): Int? {
```
Comment (yole) [inline kotlin.html.markdown]: Here and in the following two functions, obj should be Any?, otherwise the null-safety doesn't make sense

File: kotlin.html.markdown [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+ * same as `java.lang.object`.
+ *
+ * [More about classes](http://kotlinlang.org/docs/reference/classes.html)
+ */
+
+// An empty class, no curly braces needed.
+class Empty
+
+/**
+ * A class with one primary constructors and more secondary constructors. A
```
Comment (yole) [inline kotlin.html.markdown]: The class below doesn't seem to have any secondary constructors.

File: kotlin.html.markdown [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+  // The code for primary constructor should be defined in init block. If it
+  // has any parameters, they will be available.
+  init {
+    logger.info("class initialized with value ${firstName}");
+  }
+  
+  // firstName from primary constructor available in initializer.
+  val myName = firstName.toUpperCase();
+  
+  // Secondary constructors are prefixed with *constructor*.
```
Comment (yole) [inline kotlin.html.markdown]: This comment doesn't seem to be related to the class below.

File: kotlin.html.markdown (hunk 4 above)
Comment (vendethiel) [inline kotlin.html.markdown]: and "inferred" is not the correct word; the type was **computed**

Comment (loolooyyyy) [inline kotlin.html.markdown]: The official documents use "Inferred". Should "computed" be used instead?

File: kotlin.html.markdown (hunk 8 above)
Comment (loolooyyyy) [inline kotlin.html.markdown]: yes yes of course! missed it

File: kotlin.html.markdown (hunk 4 above)
Comment (vendethiel) [inline kotlin.html.markdown]: If the documentation used "inferred", it's probably better as-is, but it's not what's actually called type inference

**Comment by adambard** [comment]
We are perfectly happy to leave your PR open here until it's ready, it gives people a chance to do some review.

**Comment by loolooyyyy** [comment]
Should I follow the java doc, and add everything about basics? such as arithmetic, comparison operators and... or I keep it simple?

File: kotlin.html.markdown [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+  init {
+    logger.info("class initialized with value ${firstName}")
+  }
+  
+  // firstName from primary constructor available in initializer.
+  val myName = firstName.toUpperCase()
+  
+  // Secondary constructors are prefixed with *constructor*.
+  // Calling other constructors with *this* keyword.
+  constructor(val firstName: String, val lastName) this(firstName) {
```
Comment (yole) [inline kotlin.html.markdown]: Only primary constructor parameters can be marked as 'var' or 'val'. Hint: edit the code part of your file in IntelliJ IDEA and make sure it doesn't highlight any errors or warnings.

**Comment by yole** [comment]
I think that documenting features which are specific to Kotlin and are currently not covered in the doc (such as extension functions and function types) would be more useful and interesting than documenting basic arithmetic operators.

**Comment by loolooyyyy** [comment]
What do you think about http://kotlinlang.org/docs/kotlin-docs.pdf ?
when converted to text, it's about 3000 lines, though it may be shortened.

**Comment by levibostian** [comment]
Having a copy of the language docs is not the best method for a doc on the site. 

LXIYM is meant to be a site for existing experienced developers to quickly pick up a new language with its syntax and patterns or to use as a reference for a language they are already comfortable with. So we do not like the idea of having a language doc on the site. Not the best reference.

**Comment by levibostian** [comment]
After a quick glance, I honestly don't see much wrong with the original document they content and the way it is laid out. It reminds me of existing docs on the site which is good.

**Comment by loolooyyyy** [comment]
I'm following http://kotlinlang.org/docs/kotlin-docs.pdf but it's getting a little bit long. I'm at page 40 (of 80) and the tutorial is 600 lines long already. Feel free to remove any parts you think unnecessary.

**Comment by levibostian** [comment]
Don't worry about that yet, @loolooyyyy. Keep going. When I read docs of a new language and make a LXIYM doc it is always super long while I go through it then at the very end I read through it all and try to shorten things up/refactor/clean up so just keep going on it.

**Comment by levibostian** [comment]
Pretty excited about this, thanks a ton for doing it!

**Comment by cy6erGn0m** [comment]
@loolooyyyy Note changes in M12

**Comment by vendethiel** [comment]
@loolooyyyy any updates? :)

**Comment by vendethiel** [comment]
Going to close this one for now, since we have a kotlin doc now. Thanks!
